Issue/5460 simple payments cardreader take2

nbradbury opened the pull request:
This PR implements taking a card payment from the simple payments flow. In doing this I had three options:

1. Rewrite the card reader flow so it could be more easily shared from multiple navigation graphs
2. Duplicate the card reader flow and adapt it for simple payments
3. Have the simple payments flow show order detail and tell it to launch the card reader flow

I ruled out the first option because it's a large task and we're on a tight schedule. I originally chose the third option and implemented it [here](https://github.com/woocommerce/woocommerce-android/pull/5623), but I decided that was a terrible user experience. So in this PR I implemented the second option, despite not liking the code duplication, using a separate navigation graph for the simple payments flow.

To test:

* Go to the order list and create a simple payment
* Tap Take Payment
* Choose Card payment
* Verify the payment flow works as expected

It would also be good to verify that the card reader flows in the settings and order detail still work (they worked fine for me).

Only one reviewer is needed, but it might be good if @malinajirka or @kidinov took a quick look since they're the most familiar with the card reader codebase.

nbradbury committed:
First try at separate nav graph for simple card payment

nbradbury committed:
First attempt at replacing navigation graph with payment flow

nbradbury committed:
2nd attempt at replacing navigation graph with payment flow

nbradbury committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-android into nick/simple-payments-cardreader_take2

nbradbury committed:
Pass a bundle to the simple payments card reader flow

nbradbury committed:
Include a separate simple card reader graph

nbradbury committed:
Moved payment dialog to main nav graph

nbradbury committed:
Renamed nav graph

nbradbury committed:
Exit after payment succeeds

nbradbury committed:
Use correct included gtaph name, show progress while marking order complete

nbradbury committed:
Added comment to nav graph, corrected minor ktlint issue

nbradbury committed:
Removed unused job

nbradbury committed:
Moved the cardReaderManager to the viewModel

nbradbury committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-android into issue/5460-simple-payments-cardreader_take2

nbradbury committed:
Created separate nav graph for simple payments

nbradbury committed:
Removed simple payments out of the main nav graph

nbradbury committed:
Added global simple payment destination

nbradbury committed:
Moved showing simple payments fragment out of order list and into simple payments dialog

nbradbury committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-android into issue/5460-simple-payments-cardreader_take2

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/TakePaymentFragment.kt:
```diff
@@ -51,11 +66,36 @@ class TakePaymentFragment : BaseFragment(R.layout.fragment_take_payment) {
                     }
+                    is OrderNavigationTarget.StartCardReaderConnectFlow -> {
+                        val action = TakePaymentFragmentDirections.actionTakePaymentFragmentToCardReaderConnectDialog(
+                            skipOnboarding = true
```
🔍 Nitpick: I think we can load the `skipOnboarding` value from the `event` => `event.skipOnboarding`. Hardcoding it in the view layer could bite us. Wdyt?

nbradbury replied:
I was torn about this. Since this is "simple" payments I thought it made sense to always skip the onboarding, but perhaps you're right that I should rely on the event for this.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/TakePaymentFragment.kt:
```diff
@@ -51,11 +66,36 @@ class TakePaymentFragment : BaseFragment(R.layout.fragment_take_payment) {
+                    }
+                    is OrderNavigationTarget.StartCardReaderPaymentFlow -> {
+                        val action = TakePaymentFragmentDirections.actionTakePaymentFragmentToCardReaderPaymentDialog(
+                            orderId = viewModel.order.id
```
🔍 Similar as above, `StartCardReaderPaymentFlow` already contains `orderId`, so I think we can load it from there instead of the viewModel's field => `event.orderId`. Wdyt?

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/simplepayments/TakePaymentViewModel.kt:
```diff
@@ -60,7 +76,38 @@ class TakePaymentViewModel @Inject constructor(
+        viewState = viewState.copy(isProgressShowing = true)
+
+        launch {
+            markOrderCompleted()
```
❓ Is this really necessary? The backend updates the order state automatically and the payment flow automatically fetches the latest order from the backend. I'm a bit worried that predicting&hardcoding the final status and updating it manually on the client could easily go wrong. Wdyt?

nbradbury replied:
Oh, I didn't realize the backend marked the order complete! That certainly simplifies things. I'll remove this code.

malinajirka reviewed (commented):
Thanks for working on this @nbradbury. I'll leave the full review up to @wzieba and @ThomazFB. I've scanned through the code and left a couple nitpicks and one question.

One last thing we might want to consider is extracting `vm.onCardPaymentClicked and vm. onConnectToReaderResultReceived and vm. onCardReaderPaymentCompleted` into a separate class and re-using it. The duplication in the view layer is imo fine.

nbradbury committed:
Rely on the event rather than the view model

nbradbury committed:
Don't mark order complete after card payment completed (backend already does that)

nbradbury committed:
Fixed UnusedPrivateMember ktlint error

nbradbury committed:
Added delay to exit event

nbradbury committed:
Added snackbar on successful card payment

nbradbury commented:
@malinajirka Thanks for taking the time to look this over! I've addressed the issues you mentioned, but I've chosen to leave the duplication in the vm.

nbradbury commented:
@wzieba @ThomazFB Ready for a review!

ThomazFB reviewed (approved):
I mostly tested the feature itself since Jirka reviewed the code already, and looks good! Nice work, @nbradbury!
